File: app/src/main/java/com/trapero/cchoice/viewmodels/LoginViewModel.java
package com.trapero.cchoice.viewmodels;

import android.util.Log;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.trapero.cchoice.api.ApiService;
import com.trapero.cchoice.api.RetrofitClient;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class LoginViewModel extends ViewModel {

    private MutableLiveData<String> email = new MutableLiveData<>();
    private MutableLiveData<String> password = new MutableLiveData<>();

    private MutableLiveData<Boolean> navigateToRegister = new MutableLiveData<>(false);
    private MutableLiveData<Boolean> loginSuccess = new MutableLiveData<>(false);
    private MutableLiveData<String> errorMessage = new MutableLiveData<>();
    private MutableLiveData<Boolean> isLoggingIn = new MutableLiveData<>(false);
    private MutableLiveData<Boolean> navigateToForgotPassword = new MutableLiveData<>(false);

    public LiveData<String> getEmail() {
        return email;
    }

    public LiveData<String> getPassword() {
        return password;
    }

    public LiveData<Boolean> getNavigateToRegister() {
        return navigateToRegister;
    }

    public LiveData<Boolean> getLoginSuccess() {
        return loginSuccess;
    }

    public LiveData<String> getErrorMessage() {
        return errorMessage;
    }

    public LiveData<Boolean> getIsLoggingIn() {
        return isLoggingIn;
    }

    public LiveData<Boolean> getNavigateToForgotPassword() {
        return navigateToForgotPassword;
    }

    public void onLoginClicked() {
        String emailValue = email.getValue();
        String passwordValue = password.getValue();

        if (emailValue == null || emailValue.trim().isEmpty() || passwordValue == null || passwordValue.isEmpty()) {
            errorMessage.setValue("Please enter email and password.");
            return;
        }

        isLoggingIn.setValue(true);

        // TEMPORARILY DISABLE API CALL AND SIMULATE SUCCESS
        Log.d("LoginViewModel", "Login API call temporarily disabled. Simulating success.");
        // Simulate a successful login after a short delay (optional, for visual feedback)
        new android.os.Handler().postDelayed(() -> {
            isLoggingIn.setValue(false);
            loginSuccess.setValue(true);
        }, 1000); // Simulate a 1-second delay

        JsonObject loginData = new JsonObject();
        loginData.addProperty("email", emailValue);
        loginData.addProperty("password", passwordValue);

        ApiService apiService = RetrofitClient.getApiService();
        Call<JsonObject> call = apiService.loginUser(loginData);
        call.enqueue(new Callback<>() {
            @Override
            public void onResponse(Call<JsonObject> call, Response<JsonObject> response) {
                isLoggingIn.setValue(false);
                Log.d("LoginViewModel", "Login API Response Code: " + response.code());
                Log.d("LoginViewModel", "Login API Response Raw: " + response.toString());

                if (response.isSuccessful() && response.body() != null) {
                    JsonObject responseBody = response.body();
                    Log.d("LoginViewModel", "Login API Response Body: " + responseBody.toString());

                    if (responseBody.has("success") && responseBody.get("success").getAsBoolean()) {
                        Log.d("LoginViewModel", "Login successful according to API response.");
                        loginSuccess.setValue(true);
                        if (responseBody.has("token")) {
                            String token = responseBody.get("token").getAsString();
                            Log.d("LoginViewModel", "Login successful, token: " + token);
                            // TODO: Store the token securely
                        } else {
                            Log.w("LoginViewModel", "Login successful, but no token received.");
                        }
                        if (responseBody.has("user")) {
                            Log.d("LoginViewModel", "User data received: " + responseBody.getAsJsonObject("user").toString());
                            // TODO: Store or process user data
                        }
                    } else {
                        Log.w("LoginViewModel", "Login failed according to API response.");
                        String message = "Login failed";
                        if (responseBody.has("message")) {
                            message = "Login failed: " + responseBody.get("message").getAsString();
                        } else if (responseBody.has("error")) {
                            message = "Login failed: " + responseBody.get("error").getAsString();
                        }
                        errorMessage.setValue(message);
                        Log.e("LoginViewModel", "Login API returned failure: " + responseBody.toString());
                    }
                } else {
                    Log.e("LoginViewModel", "Login HTTP error response. Code: " + response.code() + ", Message: " + response.message());
                    String errorMessageValue = "Login failed due to a server error.";
                    if (response.errorBody() != null) {
                        try {
                            String errorString = response.errorBody().string();
                            Log.e("LoginViewModel", "Login HTTP error body: " + errorString);
                            try {
                                JsonObject errorJson = JsonParser.parseString(errorString).getAsJsonObject();
                                if (errorJson.has("message")) {
                                    errorMessageValue = "Login failed: " + errorJson.get("message").getAsString();
                                } else if (errorJson.has("error")) {
                                    errorMessageValue = "Login failed: " + errorJson.get("error").getAsString();
                                } else {
                                    errorMessageValue = "Login failed: " + errorString;
                                }
                            } catch (Exception e) {
                                errorMessageValue = "Login failed: Could not parse error response.";
                                Log.e("LoginViewModel", "Error parsing error body: " + e.getMessage());
                            }
                        } catch (Exception e) {
                            errorMessageValue = "Login failed: Could not read error response.";
                            Log.e("LoginViewModel", "Error reading error body: " + e.getMessage());
                        }
                    }
                    errorMessage.setValue(errorMessageValue);
                }
            }

            @Override
            public void onFailure(Call<JsonObject> call, Throwable t) {
                isLoggingIn.setValue(false);
                errorMessage.setValue("Network error: " + t.getMessage());
                Log.e("LoginViewModel", "Login network error: " + t.getMessage());
            }
        });

    }

    public void onSignUpClicked() {
        navigateToRegister.setValue(true);
    }

    public void onForgotPasswordClicked() {
        navigateToForgotPassword.setValue(true);
    }

    public void doneNavigatingToRegister() {
        navigateToRegister.setValue(false);
    }

    public void doneLoggingIn() {
        loginSuccess.setValue(false);
    }

    public void doneShowingError() {
        errorMessage.setValue(null);
    }

    public void doneNavigatingToForgotPassword() {
        navigateToForgotPassword.setValue(false);
    }
}

File: app/src/main/java/com/trapero/cchoice/viewmodels/LoginViewModel.kt
package com.trapero.cchoice.viewmodels

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.google.gson.Gson
import com.google.gson.JsonObject
import com.google.gson.reflect.TypeToken
import com.trapero.cchoice.api.ApiHelper
import com.trapero.cchoice.session.Session
import kotlin.Unit
import kotlin.jvm.functions.Function1
import okhttp3.Call
import okhttp3.Callback
import okhttp3.Response
import java.io.IOException

class LoginViewModel : ViewModel() {

    // Using correct setter with backing field
    private val _email = MutableLiveData<String>()
    var email: MutableLiveData<String>
        get() = _email
        set(value) { _email.value = value.value }

    private val _password = MutableLiveData<String>()
    var password: MutableLiveData<String>
        get() = _password
        set(value) { _password.value = value.value }

    private val _navigateToRegister = MutableLiveData<Boolean>(false)
    var navigateToRegister: MutableLiveData<Boolean>
        get() = _navigateToRegister
        set(value) { _navigateToRegister.value = value.value }

    private val _loginSuccess = MutableLiveData<Boolean>(false)
    var loginSuccess: MutableLiveData<Boolean>
        get() = _loginSuccess
        set(value) { _loginSuccess.value = value.value }

    private val _errorMessage = MutableLiveData<String?>()
    var errorMessage: MutableLiveData<String?>
        get() = _errorMessage
        set(value) { _errorMessage.value = value.value }

    private val _isLoggingIn = MutableLiveData<Boolean>(false)
    var isLoggingIn: MutableLiveData<Boolean>
        get() = _isLoggingIn
        set(value) { _isLoggingIn.value = value.value }

    private val _navigateToForgotPassword = MutableLiveData<Boolean>(false)
    var navigateToForgotPassword: MutableLiveData<Boolean>
        get() = _navigateToForgotPassword
        set(value) { _navigateToForgotPassword.value = value.value }

    fun onLoginClicked() {
        val emailValue = email.value
        val passwordValue = password.value

        isLoggingIn.value = true

        val apiHelper = ApiHelper()

        apiHelper.setBody <JsonObject>( mapOf<String,Any>(
            "user_email" to email.value.toString(),
            "user_password" to password.value.toString()
        ))
        .post("/login", object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                isLoggingIn.postValue(false)
                errorMessage.postValue(e.message)
            }

            override fun onResponse(call: Call, response: Response) {
                isLoggingIn.postValue(false)
                if (response.isSuccessful) {
                    loginSuccess.postValue(true)
                    val mapType = object : TypeToken<Map<String, Any>>() {}.type
                    val responseMap: Map<String, Any> = Gson().fromJson(response.body?.string(), mapType)
                    val token = responseMap["token"] as? String
                    token?.let { Session.authToken = it }
                } else {
                    val mapType = object : TypeToken<Map<String, Any>>() {}.type
                    val responseMap: Map<String, Any> = Gson().fromJson(response.body?.string(), mapType)
                    Log.i("tesstKim", responseMap.toString())
                    val errors = responseMap["errors"] as? Map<String, ArrayList<String>>
                    val error = responseMap["error"] as? Map<String, Any>
                    val firstErrorMessage = errors?.values?.firstOrNull()?.firstOrNull()

                    if(firstErrorMessage != null)
                        errorMessage.postValue(firstErrorMessage)
                    else if (error?.get("message") as? String != null)
                        errorMessage.postValue(error["message"].toString())
                    else
                        errorMessage.postValue("No errors specified")
                }
            }
        })
    }

    fun onSignUpClicked() {
        navigateToRegister.value = true
    }

    fun onForgotPasswordClicked() {
        navigateToForgotPassword.value = true
    }

    fun doneNavigatingToRegister() {
        navigateToRegister.value = false
    }

    fun doneLoggingIn() {
        loginSuccess.value = false
    }

    fun doneShowingError() {
        errorMessage.value = null
    }

    fun doneNavigatingToForgotPassword() {
        navigateToForgotPassword.value = false
    }
}
